### PR TITLE
MCPServer: add active unit gauge with prometheus_client

### PR DESCRIPTION
**Pull request by sevein**
Connects to https://github.com/artefactual/archivematica/issues/1187.

### CODE REVIEW

**Comment by sevein** [comment]
@marktriggs can you take a look? The Prometheus client library has a global `CollectorRegistry`, that's where the samples are kept until a Prometheus server collects them and puts them in the time series database.

I suspect we still want to do the sampling + max() that you implemented.

**Comment by sevein** [comment]
Example of document generated by the prometheus http interface:

```
# HELP task_group_runner_active_units Number of units currently being processed
# TYPE task_group_runner_active_units gauge
task_group_runner_active_units 3.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="2",minor="7",patchlevel="15",version="2.7.15"} 1.0
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 3427557376.0
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 63528960.0
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1531346625.42
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 19.240000000000002
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 14.0
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1048576.0
# HELP rpc_job_package_create_request_time Time spent processing request
# TYPE rpc_job_package_create_request_time summary
rpc_job_package_create_request_time_count 3.0
rpc_job_package_create_request_time_sum 0.026831865310668945
# HELP request_processing_seconds Time spent processing request
# TYPE request_processing_seconds summary
request_processing_seconds_count 0.0
request_processing_seconds_sum 0.0
# HELP rpc_job_awaiting_approval_request_time Time spent processing request
# TYPE rpc_job_awaiting_approval_request_time summary
rpc_job_awaiting_approval_request_time_count 108.0
rpc_job_awaiting_approval_request_time_sum 0.040329933166503906
# HELP rpc_job_approve_request_time Time spent processing request
# TYPE rpc_job_approve_request_time summary
rpc_job_approve_request_time_count 0.0
rpc_job_approve_request_time_sum 0.0
```

**Comment by marktriggs** [comment]
@sevein Cool, that looks good.  And yeah, I think the sampling is still needed.

It's going to update that gauge value 5 times per second but it looks like that won't cause problems... it seems like each update just replaces the previous value, so not chewing up memory or anything like that.

**Comment by sevein** [comment]
Here's the gauge in action using Prometheus's graphing tool. It's showing a period where ten transfers were started using the new Package endpoint. The samples captured are higher than the number of expected units at some point. I suspect this is because we have a few entities (Transfer, SIP, DIP...) related to the same package, not something that we can fix now.

image

File: src/MCPServer/lib/taskGroupRunner.py [hunk 1]
```diff
@@ -218,6 +241,12 @@ def _monitor_running_jobs(self, gm_client):
                         len(self.task_group_jobs_by_uuid))
             self.last_notification_time = now
 
+        active_count = len(set([task_group.task_group.unit_uuid
```
Comment (marktriggs) [inline src/MCPServer/lib/taskGroupRunner.py]: Oops, this is a bug from me.  I meant to write:

```
 active_count = len(set([task_group.task_group.unit_uuid()
```

(i.e. call `unit_uuid` as a method--missed the trailing parens!)

It works by accident at the moment.

**Comment by marktriggs** [comment]
@sevein The inflated numbers are interesting... with a single transfer, I can see the unit UUID switching as it moves from Transfer to SIP, but I was surprised it got counted twice since I thought those stages wouldn't overlap.

I think it might just be that the move_transfer client script has finished moving the transfer into the watched directory but hasn't yet returned its result to Gearman.  Then, the watched directory picks up the new SIP and sends the task to the task runner, which counts it.  So yeah.  Tricky :)

**Comment by sevein** [comment]
Ok, fixed. Rewrote your commit.

**Comment by marktriggs** [comment]
Much obliged, thanks :)

File: src/MCPServer/lib/archivematicaMCP.py [hunk 2]
```diff
@@ -338,6 +339,8 @@ def _except_hook_log_everything(exc_type, exc_value, exc_traceback):
     logger.info('This PID: %s', os.getpid())
     logger.info('User: %s', getpass.getuser())
 
+    start_http_server(7999)
```
Comment (sevein) [inline src/MCPServer/lib/archivematicaMCP.py]: I need to make this configurable.

**Comment by sevein** [comment]
I've added a new `PROMETHEUS_HTTP_SERVER` setting that defaults to empty string. The Prometheus server is only created when the setting carries a non-empty string, e.g. forms `12345` or `127.0.0.1:12345` are accepted - ignored otherwise.

**Pull request merged**